### PR TITLE
Integrate Caduceus Model

### PR DESCRIPTION
**Pull request by mattwoodx**
- Add the Caduceus model to the package
- This model is dependent on having a CUDA enabled GPU and mamba_ssm so it has some checks to see if these are present in order for the module to be imported.

### CODE REVIEW

File: examples/run_models/configs/caduceus_config.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+model_name: "caduceus-ph-4L-seqlen-1k-d118"
+batch_size: 2
+pooling_strategy: "mean"
+device: "cuda"
```
Comment (bputzeys) [inline examples/run_models/configs/caduceus_config.yaml:4]: If cuda GPU is a strict requirement, I would take it out here and hardcode it after the check in `Caduceaus.__init__()`. Otherwise the user might try `cpu`, which does not work

File: helical/models/caduceus/pretrained_config.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+Taken from Caduceus repository: https://github.com/kuleshov-group/caduceus/blob/main/caduceus/configuration_caduceus.py
+"""
+
+from typing import Optional, Union
+
+from transformers import PretrainedConfig
+
+
+class CaduceusPretrainedConfig(PretrainedConfig):
```
Comment (bputzeys) [inline helical/models/caduceus/pretrained_config.py:11]: docstrings

File: helical/models/caduceus/pretrained_config.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    """Config that extends the original MambaConfig with params relevant to bi-directionality and RC equivariance."""
+    def __init__(
+            self,
+            # From original MambaConfig
+            d_model: int = 2560,
```
Comment (bputzeys) [inline helical/models/caduceus/pretrained_config.py:16]: currently, for the user, there is no way to set these parameters. Can you make them available in the hydra config?

File: helical/models/caduceus/model.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            Containing processed DNA sequences.
+
+        """
+        LOGGER.info("Processing data")
+
+        self.ensure_dna_sequence_validity(sequences)
+        
+        tokenized_sequences = []
+        for seq in sequences:
+            tokenized_seq = self.tokenizer(seq, return_tensors="pt", padding="max_length", truncation=True, max_length=self.config['input_size'])
```
Comment (bputzeys) [inline helical/models/caduceus/model.py:118]: make these parameters available to be set in the function `process_data()` above so that they could be modified.

File: helical/models/caduceus/model.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            The output dataset from `process_data`.
+
+        Returns
+        ----------
+        np.ndarray
+            The embeddings for the tokenized sequence in the form of a numpy array. 
+            NOTE: This method returns the embeddings using the pooling strategy specified in the config.
+        """
+        LOGGER.info("Inference started")
+        dataloader = DataLoader(dataset, collate_fn=self._collate_fn, batch_size=self.config['batch_size'], shuffle=False, num_workers=self.config['nproc'])
```
Comment (bputzeys) [inline helical/models/caduceus/model.py:138]: make `shuffle` available as input parameter to `get_embeddings`

File: helical/models/caduceus/model.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    
+        embeddings = []
+        self.model.to(self.device)
```
Comment (bputzeys) [inline helical/models/caduceus/model.py:141]: is this necessary if it is GPU by default?

File: ci/tests/test_caduceus/test_caduceus_model.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import pytest
+import numpy as np
+import torch
+
+try:
+    from helical import Caduceus, CaduceusConfig
+    skip = False # only run tests if able to import the package
```
Comment (bputzeys) [inline ci/tests/test_caduceus/test_caduceus_model.py:7]: rename this `skip` variable

File: ci/tests/test_caduceus/test_caduceus_model.py [hunk 8]
```diff
@@ -4,12 +4,12 @@
 
 try:
     from helical import Caduceus, CaduceusConfig
-    skip = False # only run tests if able to import the package
+    caduceus_unavailable = False # only run tests if able to import the package
```
Comment (bputzeys) [inline ci/tests/test_caduceus/test_caduceus_model.py:7]: nice

File: ci/tests/test_caduceus/test_caduceus_model.py [hunk 9]
```diff
@@ -46,27 +46,41 @@ def test_runtime_error_when_cuda_unavailable(mocker):
...
+@pytest.mark.parametrize("input_sequence, expected_sequence", [
     # Valid DNA sequences
-    "A",
-    "CC",
-    "TTTT", 
-    "ACGTN",
-    "ACG" * 256
+    ("A", [7, 1]),
+    ("CC", [8, 8, 1]),
+    ("TTTT", [10, 10, 10, 10, 1]),
```
Comment (bputzeys) [inline ci/tests/test_caduceus/test_caduceus_model.py:54]: cool

File: helical/models/caduceus/caduceus_config.py [hunk 10]
```diff
@@ -83,6 +82,6 @@ def __init__(
             "model_name": model_name,
             "pooling_strategy": pooling_strategy,
             "batch_size": batch_size,   
-            "device": device,
+            "device": "cuda",
```
Comment (bputzeys) [inline helical/models/caduceus/caduceus_config.py:85]: add comment that this is a strict requirement
`# Caduceus using Triton for mamba-SSM which requires cuda`

File: helical/models/caduceus/caduceus_config.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+        Parameters
+        ----------
+        model_name : Literal["caduceus-ph-4L-seqlen-1k-d118", "caduceus-ph-4L-seqlen-1k-d256", "caduceus-ph-16L-seqlen-131k-d256", "caduceus-ps-4L-seqlen-1k-d118", "caduceus-ps-4L-seqlen-1k-d256", "caduceus-ps-16L-seqlen-131k-d256"], optional, default="caduceus-ph-4L-seqlen-1k-d118"
+            The model configuration name to use.
+        batch_size : int, optional, default=5
+            The batch size. This will be used in all the model operations.
+        pooling_strategy : Literal["mean", "max", "last", "first"], optional, default="mean"
+            The pooling strategy to use. This will be used in all the model operations.
+        device : str, optional, default="cuda"
```
Comment (bputzeys) [inline helical/models/caduceus/caduceus_config.py:27]: remove this

**Review by bputzeys — APPROVED** [review]
